File: app/src/main/java/com/paxus/pay/poslinkui/demo/viewmodel/SecondScreenInfoViewModel.java
package com.paxus.pay.poslinkui.demo.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

/**
 * Author: Elaine Xie
 * Date: 2025/12/8
 * Desc:
 */
public class SecondScreenInfoViewModel extends ViewModel {
    private final MutableLiveData<String> title = new MutableLiveData<>("");
    private final MutableLiveData<String> amount = new MutableLiveData<>("");
    private final MutableLiveData<String> msg = new MutableLiveData<>("");
    private final MutableLiveData<String> status = new MutableLiveData<>("");
    private final MutableLiveData<String> statusTitle = new MutableLiveData<>("");
    private final MutableLiveData<Integer> imageResourceId = new MutableLiveData<>();

    private final MediatorLiveData<ScreenInfo> screenInfo = new MediatorLiveData<>();


    public void updateAmount(String newContent) {
        amount.setValue(newContent);
    }

    public void updateTitle(String newContent) {
        title.setValue(newContent);
    }

    public void updateMessage(String newContent) {
        msg.setValue(newContent);
    }

    public SecondScreenInfoViewModel() {
        screenInfo.addSource(amount, amount -> combineUserData());
        screenInfo.addSource(msg, msg -> combineUserData());
        screenInfo.addSource(status, status -> combineUserData());
        screenInfo.addSource(imageResourceId, status -> combineUserData());
        screenInfo.addSource(title, status -> combineUserData());
        screenInfo.addSource(statusTitle, status -> combineUserData());
    }

    private void combineUserData() {
        String newAmount = amount.getValue();
        String newMsg = msg.getValue();
        String newStatus = status.getValue();
        Integer newImageResourceId = imageResourceId.getValue();
        String newTitle = title.getValue();
        String newStatusTitle = statusTitle.getValue();

        ScreenInfo info = new ScreenInfo(newAmount, newMsg, newStatus, newImageResourceId, newTitle, newStatusTitle);
        screenInfo.setValue(info);
    }

    public LiveData<ScreenInfo> getScreenInfo() {
        return screenInfo;
    }

    //  update second screen info
    public void updateAllData(String newAmount, String newMsg, String newStatus, Integer newImageResourceId, String newStatusTitle, String newTitle) {
        amount.setValue(newAmount);
        msg.setValue(newMsg);
        status.setValue(newStatus);
        imageResourceId.setValue(newImageResourceId);
        statusTitle.setValue(newStatusTitle);
        title.setValue(newTitle);
    }

    public static class ScreenInfo {
        public final String title;
        public final String amount;
        public final String msg;
        public final String status;
        public final String statusTitle;
        public final Integer imageResourceId;

        public ScreenInfo(String amount, String msg, String status, Integer imageResourceId, String title, String statusTitle) {
            this.amount = amount;
            this.msg = msg;
            this.status = status;
            this.imageResourceId = imageResourceId;
            this.title = title;
            this.statusTitle = statusTitle;
        }
    }
}

File: app/src/main/java/com/paxus/pay/poslinkui/demo/viewmodel/SecondScreenInfoViewModel.kt
package com.paxus.pay.poslinkui.demo.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn

/**
 * Second-screen line items for the demo; exposes aggregated [screenInfo] as [StateFlow] (no LiveData).
 *
 * Uses nested [combine] because the Flow API provides overloads only up to five flows.
 */
@HiltViewModel
class SecondScreenInfoViewModel @Inject constructor() : ViewModel() {
    private val title = MutableStateFlow<String?>("")
    private val amount = MutableStateFlow<String?>("")
    private val msg = MutableStateFlow<String?>("")
    private val status = MutableStateFlow<String?>("")
    private val statusTitle = MutableStateFlow<String?>("")
    private val imageResourceId = MutableStateFlow<Int?>(null)

    // kotlinx.coroutines.flow.combine has overloads up to 5 flows; nest for six inputs.
    val screenInfo: StateFlow<ScreenInfo?> = combine(
        combine(amount, msg, status) { am, m, st -> Triple(am, m, st) },
        combine(imageResourceId, title, statusTitle) { img, ti, stt -> Triple(img, ti, stt) }
    ) { amountMsgStatus, imageTitleStatusTitle ->
        buildScreenInfo(
            amountMsgStatus.first,
            amountMsgStatus.second,
            amountMsgStatus.third,
            imageTitleStatusTitle.first,
            imageTitleStatusTitle.second,
            imageTitleStatusTitle.third
        )
    }.stateIn(
        viewModelScope,
        SharingStarted.WhileSubscribed(5_000),
        buildScreenInfo("", "", "", null, "", "")
    )

    fun updateAmount(newContent: String?) {
        amount.value = newContent
    }

    fun updateTitle(newContent: String?) {
        title.value = newContent
    }

    fun updateMessage(newContent: String?) {
        msg.value = newContent
    }

    internal fun buildScreenInfo(
        amount: String?,
        msg: String?,
        status: String?,
        imageResourceId: Int?,
        title: String?,
        statusTitle: String?
    ): ScreenInfo {
        return ScreenInfo(amount, msg, status, imageResourceId, title, statusTitle)
    }

    fun updateAllData(
        newAmount: String?,
        newMsg: String?,
        newStatus: String?,
        newImageResourceId: Int?,
        newStatusTitle: String?,
        newTitle: String?
    ) {
        amount.value = newAmount
        msg.value = newMsg
        status.value = newStatus
        imageResourceId.value = newImageResourceId
        statusTitle.value = newStatusTitle
        title.value = newTitle
    }

    class ScreenInfo(
        val amount: String?,
        val msg: String?,
        val status: String?,
        val imageResourceId: Int?,
        val title: String?,
        val statusTitle: String?
    )
}
